agents/supervisor.py

```diff
@@ -0,0 +1,179 @@
+import logging
+
+import coro
+
+from .utils.logtools import maybe_log_message
+
+
+class AgentSupervisor(object):
+    """
+    Supervisor that monitors and manages agent coroutines.
+
+    Attributes:
+        agent (ServerAgent): Agent instance under supervision.
+        coros (dict): Dictionary mapping coroutine IDs to coroutine
+            references. ID 0 is reserved for exit coroutine.
+
+    Methods:
+        last_coro(): Get ID of the last scheduled coroutine.
+        start(): Start the event loop and block until explicitly stopped.
+        sleep(interval): Yield to event loop and sleep for a duration of
+            the interval.
+        schedule(task, max_retries, interval, idx, ...): Schedule a
+            periodic coroutine task.
+        unschedule(idx): Unschedule a coroutine by ID.
+        schedule_exit(interval, prestop, ...): Schedule a periodic
+            coroutine to monitor for exit.
+    """
+
+    def __init__(self, agent):
+        self.agent = agent
+
+        self.coros = {}  # Store coroutine IDs and references
```

Comment:
Consider guarding `coros` access with a lock or atomic operations if there’s any chance of multithreading in the future. Even though coro is cooperative, safety helps in long-lived agents.

Reply:
Done this using `threading.Lock`. It's really convenient to use dict here, and I didn't find a better option with a dict.

Working with `coro` then goes like this:

```python
def put_coro(self, idx, coroutine):
    with self._lock:
        self._coros[idx] = coroutine
```